tree-diff

```diff
@@ -0,0 +1,156 @@
+                         os.path.join(f"/proc/self/fd/{dir_fd2}", dirent.name),
+                         os.path.join(path, dirent.name),
+                         report["differences"])
+            if not stat_diff(stat1,
+                             stat2,
+                             os.path.join(path, dirent.name),
+                             report["differences"]):
+                continue
+            if dirent.is_symlink():
+                symlink_diff(dirent.name,
+                             dir_fd1,
+                             dir_fd2,
+                             os.path.join(path, dirent.name),
+                             report["differences"])
+            elif dirent.is_file(follow_symlinks=False):
+                content_diff(dirent.name,
+                             dir_fd1,
+                             dir_fd2,
+                             stat1.st_size,
+                             stat2.st_size,
+                             os.path.join(path, dirent.name),
+                             report["differences"])
+            elif dirent.is_dir(follow_symlinks=False):
+                try:
+                    child_fd1 = os.open(dirent.name, os.O_DIRECTORY, dir_fd=dir_fd1)
+                except OSError:
+                    continue
+                try:
+                    child_fd2 = os.open(dirent.name, os.O_DIRECTORY, dir_fd=dir_fd2)
+                except OSError:
+                    os.close(child_fd1)
+                    continue
+                diff_aux(child_fd1, child_fd2, os.path.join(path, dirent.name), report)
+                os.close(child_fd2)
+                os.close(child_fd1)
+    with os.scandir(dir_fd2) as it:
+        for dirent in it:
+            if dirent.name not in entries1:
+                report["added_files"] += [os.path.join(path, dirent.name)]
+
+
+def diff(dir_fd1, dir_fd2, report):
+    stat1 = os.stat(".", dir_fd=dir_fd1, follow_symlinks=False)
+    stat2 = os.stat(".", dir_fd=dir_fd2, follow_symlinks=False)
+    selinux_diff(f"/proc/self/fd/{dir_fd1}", f"/proc/self/fd/{dir_fd2}", "/", report["differences"])
+    stat_diff(stat1, stat2, "/", report["differences"])
+    diff_aux(dir_fd1, dir_fd2, "/", report)
+
+
+def main():
+    parser = argparse.ArgumentParser(description="Recursively compare file system trees")
+    parser.add_argument("dir1", metavar="DIRECTORY1",
+                        help="first directory to compare")
+    parser.add_argument("dir2", metavar="DIRECTORY2",
+                        help="second directory to compare")
+    args = parser.parse_args()
+
+    report = {}
+    report["added_files"] = []
+    report["deleted_files"] = []
+    report["differences"] = {}
+
+    dir_fd1 = os.open(args.dir1, os.O_DIRECTORY)
+    dir_fd2 = os.open(args.dir2, os.O_DIRECTORY)
+    diff(dir_fd1, dir_fd2, report)
+    os.close(dir_fd2)
+    os.close(dir_fd1)
+
+    print(json.dumps(report, indent=2))
```

Comment:
nitpick: I'd have used `json.dump(report, sys.stdout, indent=2)` here. Doesn't really matter of course.